File: ios/Sublight/RoverImages.swift
import Foundation

// Live, on-demand access to a rover's raw images for a single sol.
// The full archive is enormous, so nothing is bundled; we fetch a sample of a
// sol's frames from the public NASA feeds when the user taps a drive stop.

// mars.nasa.gov serves size variants of every MSL frame by suffix, and the
// saving is dramatic: a NAVCAM frame is 536 KB raw, 113 KB as -br, 8 KB as
// -thm. Feeding the raw frame into a 34pt filmstrip cell was what made images
// feel slow to arrive. Verified present on every MSL camera.
func mslVariant(_ url: URL, _ suffix: String) -> URL {
    let s = url.absoluteString
    guard let dot = s.range(of: ".", options: .backwards), dot.lowerBound > s.startIndex else { return url }
    let ext = String(s[dot.lowerBound...])
    guard ext.lowercased() == ".jpg" else { return url }
    return URL(string: String(s[s.startIndex..<dot.lowerBound]) + suffix + ext) ?? url
}

struct RoverImage: Identifiable, Sendable {
    let id = UUID()
    /// Smallest size: the scrub strip.
    let thumb: URL
    /// Mid size for the feed stage.
    let view: URL
    let full: URL
    let sourceUrl: URL
    let instrument: String
    let capturedUtc: String
    /// When the frame actually reached Earth: the feed's `date_received`. This
    /// is what the gallery orders and dates by. Capture time plus light-time is
    /// not an arrival — a rover buffers frames and downlinks them through a
    /// relay orbiter hours or days after the shutter.
    let receivedUtc: String
    let sol: Int
}

/// Rovers shoot in stereo: the same instant through a left and a right eye, and
/// both eyes are published as separate frames. A sol of 214 Perseverance frames
/// is really about 149 scenes. Keep the left eye when a right one shares its
/// capture time and its camera differs only by which eye it is.
///
/// Measured live: Perseverance sol 1972 goes 214 to 149, Curiosity sol 5008
/// goes 199 to 111.
func dropStereoTwins(_ frames: [RoverImage]) -> [RoverImage] {
    var seen: [String: Int] = [:]
    var out: [RoverImage] = []
    for f in frames {
        let base = f.instrument
            .replacingOccurrences(of: "_LEFT", with: "")
            .replacingOccurrences(of: "_RIGHT", with: "")
        let key = "\(f.capturedUtc)|\(base)"
        guard let at = seen[key] else {
            seen[key] = out.count
            out.append(f)
            continue
        }
        if f.instrument.contains("LEFT") && out[at].instrument.contains("RIGHT") { out[at] = f }
    }
    return out
}

/// How much a camera is worth opening a post on, lowest number first. The raw
/// feeds carry no "featured" or "interesting" flag; the closest thing NASA
/// publishes is MSL's `instrument_sort` (Mastcam 1, ChemCam RMI 4, Navcam 7-8),
/// and this matches it while also covering Perseverance, which has no
/// equivalent field.
func cameraRank(_ instrument: String) -> Int {
    if instrument.contains("MCZ") || instrument.contains("MAST") || instrument.contains("ZCAM") { return 0 }
    if instrument.contains("NAVCAM") || instrument.contains("NAV_") { return 1 }
    if instrument.contains("RMI") || instrument.contains("SUPERCAM") || instrument.contains("CHEMCAM") { return 2 }
    if instrument.contains("HAZ") { return 3 }
    return 4
}

struct SolImages: Sendable {
    let sol: Int
    let count: Int
    let images: [RoverImage]
    let moreURL: URL?
}

/// The caches are touched by several tasks at once: the feed's own refresh, a
/// pull to refresh, a return to the foreground, and the background check behind
/// notifications. A Swift Dictionary is not safe under concurrent writes, and
/// this crashed the app on launch with a segfault inside
/// `Dictionary._Variant.setValue`. An actor serialises every access.
private actor ImageCache {
    private var sols: [String: SolImages] = [:]
    private var latest: [String: (at: Date, images: [RoverImage])] = [:]

    func sol(_ key: String) -> SolImages? { sols[key] }
    func store(_ key: String, _ value: SolImages) { sols[key] = value }

    func latestFresh(_ key: String, ttl: TimeInterval) -> [RoverImage]? {
        guard let hit = latest[key], Date().timeIntervalSince(hit.at) < ttl else { return nil }
        return hit.images
    }
    func storeLatest(_ key: String, _ images: [RoverImage]) { latest[key] = (Date(), images) }
}

enum RoverImages {
    private static let store = ImageCache()

    // ---- newest published frames --------------------------------------------
    // The bundled snapshot is only as fresh as the last data refresh, and NASA
    // publishes in bursts through the day, so the feed asks for the most
    // recently *published* frames (ordered by date_received) rather than
    // guessing a sol.
    private static let latestTTL: TimeInterval = 300

    /// `force` skips the cache on the way in (never on the way out). Re-opening
    /// the gallery, pulling to refresh and the background check all take this
    /// path: the sol on screen is still being added to, so a cached answer is
    /// the wrong answer.
    static func fetchLatest(roverId: String, limit: Int = 48, force: Bool = false) async -> [RoverImage] {
        let key = "\(roverId):\(limit)"
        if !force, let hit = await store.latestFresh(key, ttl: latestTTL) { return hit }
        let images: [RoverImage]
        if roverId == "curiosity" {
            images = (try? await latestCuriosity(limit: limit)) ?? []
        } else {
            images = (try? await latestPerseverance(limit: limit)) ?? []
        }
        let deduped = dropStereoTwins(images)
        if !deduped.isEmpty { await store.storeLatest(key, deduped) }
        return deduped
    }

    private static func latestPerseverance(limit: Int) async throws -> [RoverImage] {
        let urlStr = "https://mars.nasa.gov/rss/api/?feed=raw_images&category=mars2020&feedtype=json"
            + "&num=\(limit)&page=0&order=date_received+desc"
        let (data, _) = try await URLSession.shared.data(from: URL(string: urlStr)!)
        let d = try JSONDecoder().decode(M20Response.self, from: data)
        return d.images.compactMap { im in
            let f = im.image_files
            guard let thumbS = f?.small ?? f?.medium ?? f?.large ?? f?.full_res,
                  let thumb = URL(string: thumbS) else { return nil }
            let fullS = f?.large ?? f?.full_res ?? f?.medium ?? thumbS
            let viewS = f?.medium ?? f?.large ?? thumbS
            return RoverImage(thumb: thumb, view: URL(string: viewS) ?? thumb, full: URL(string: fullS) ?? thumb,
                              sourceUrl: URL(string: im.link ?? fullS) ?? thumb,
                              instrument: im.camera?.instrument ?? "CAMERA",
                              capturedUtc: im.date_taken_utc ?? "",
                              receivedUtc: im.date_received ?? "", sol: im.sol ?? 0)
        }
    }

    private static func latestCuriosity(limit: Int) async throws -> [RoverImage] {
        let urlStr = "https://mars.nasa.gov/api/v1/raw_image_items/?order=date_received+desc"
            + "&per_page=\(limit * 2)&page=0&condition_1=msl%3Amission"
        let (data, _) = try await URLSession.shared.data(from: URL(string: urlStr)!)
        let d = try JSONDecoder().decode(MSLResponse.self, from: data)
        return d.items.filter { $0.is_thumbnail != true }.prefix(limit).compactMap { im in
            guard let s = im.url, let u = URL(string: s) else { return nil }
            return RoverImage(thumb: mslVariant(u, "-thm"), view: mslVariant(u, "-br"), full: u, sourceUrl: u,
                              instrument: im.instrument ?? "CAMERA",
                              capturedUtc: im.date_taken ?? "",
                              receivedUtc: im.date_received ?? "", sol: im.sol ?? 0)
        }
    }

    static func fetch(roverId: String, sol: Int, limit: Int = 120, force: Bool = false) async -> SolImages {
        let key = "\(roverId):\(sol):\(limit)"
        if !force, let hit = await store.sol(key) { return hit }
        let result: SolImages
        if roverId == "curiosity" {
            result = (try? await fetchCuriosity(sol: sol, limit: limit)) ?? SolImages(sol: sol, count: 0, images: [], moreURL: nil)
        } else {
            result = (try? await fetchPerseverance(sol: sol, limit: limit)) ?? SolImages(sol: sol, count: 0, images: [], moreURL: nil)
        }
        await store.store(key, result)
        return result
    }

    // MARK: - Perseverance (mars2020 raw images)

    private struct M20Response: Decodable {
        struct Image: Decodable {
            struct Files: Decodable { let small: String?; let medium: String?; let large: String?; let full_res: String? }
            struct Camera: Decodable { let instrument: String? }
            let image_files: Files?
            let camera: Camera?
            let date_taken_utc: String?
            let date_received: String?
            let sol: Int?
            let link: String?
        }
        let images: [Image]
        let num_images: Int?
    }

    private static func fetchPerseverance(sol: Int, limit: Int) async throws -> SolImages {
        let urlStr = "https://mars.nasa.gov/rss/api/?feed=raw_images&category=mars2020&feedtype=json"
            + "&num=\(limit)&page=0&order=sol+desc&sol=\(sol)"
        let (data, _) = try await URLSession.shared.data(from: URL(string: urlStr)!)
        let d = try JSONDecoder().decode(M20Response.self, from: data)
        let images: [RoverImage] = d.images.prefix(limit).compactMap { im in
            let f = im.image_files
            guard let thumbS = f?.small ?? f?.medium ?? f?.large ?? f?.full_res,
                  let thumb = URL(string: thumbS) else { return nil }
            let fullS = f?.large ?? f?.full_res ?? f?.medium ?? thumbS
            let viewS = f?.medium ?? f?.large ?? thumbS
            return RoverImage(
                thumb: thumb,
                view: URL(string: viewS) ?? thumb,
                full: URL(string: fullS) ?? thumb,
                sourceUrl: URL(string: im.link ?? fullS) ?? thumb,
                instrument: im.camera?.instrument ?? "CAMERA",
                capturedUtc: im.date_taken_utc ?? "",
                receivedUtc: im.date_received ?? "",
                sol: im.sol ?? sol)
        }
        let more = URL(string: "https://mars.nasa.gov/mars2020/multimedia/raw-images/?order=sol+desc&per_page=100&page=0&begin_sol=\(sol)&end_sol=\(sol)")
        return SolImages(sol: sol, count: d.num_images ?? images.count, images: dropStereoTwins(images), moreURL: more)
    }

    // MARK: - Curiosity (msl raw image items)

    private struct MSLResponse: Decodable {
        struct Item: Decodable {
            let url: String?
            let instrument: String?
            let date_taken: String?
            let date_received: String?
            let sol: Int?
            let is_thumbnail: Bool?
        }
        let items: [Item]
        let total: Int?
    }

    private static func fetchCuriosity(sol: Int, limit: Int) async throws -> SolImages {
        // Each full frame ships with a low-res `is_thumbnail` twin, so over-fetch
        // and drop the thumbnails before trimming to `limit` full frames.
        let urlStr = "https://mars.nasa.gov/api/v1/raw_image_items/?order=sol+desc&per_page=\(limit * 2)&page=0"
            + "&condition_1=msl%3Amission&condition_2=\(sol)%3Asol%3Agte&condition_3=\(sol)%3Asol%3Alte"
        let (data, _) = try await URLSession.shared.data(from: URL(string: urlStr)!)
        let d = try JSONDecoder().decode(MSLResponse.self, from: data)
        let full = d.items.filter { $0.is_thumbnail != true }
        let images: [RoverImage] = full.prefix(limit).compactMap { im in
            guard let s = im.url, let u = URL(string: s) else { return nil }
            return RoverImage(thumb: mslVariant(u, "-thm"), view: mslVariant(u, "-br"), full: u, sourceUrl: u,
                              instrument: im.instrument ?? "CAMERA",
                              capturedUtc: im.date_taken ?? "",
                              receivedUtc: im.date_received ?? "", sol: im.sol ?? sol)
        }
        let more = URL(string: "https://mars.nasa.gov/msl/multimedia/raw-images/?order=sol+desc&per_page=100&page=0&begin_sol=\(sol)&end_sol=\(sol)")
        return SolImages(sol: sol, count: full.count, images: dropStereoTwins(images), moreURL: more)
    }
}

File: src/data/roverImages.ts
// Live, on-demand access to a rover's raw images for a single sol.
//
// The full archive is enormous (Perseverance alone has >560,000 raw frames), so
// nothing is bundled. Both feeds send `Access-Control-Allow-Origin: *`, so — like
// the DSN feed — the browser reads them directly when the user opens a drive stop.
// Every frame stays linked; none is hosted or invented.
import type { FrameThumb } from '../types.ts';

// mars.nasa.gov serves size variants of every MSL frame by suffix, and the
// saving is dramatic: a NAVCAM frame is 536 KB raw, 113 KB as -br, 8 KB as
// -thm. Serving the raw frame into a 34px filmstrip cell was what made the
// feed feel slow. Verified present on every MSL camera.
const mslVariant = (url: string, suffix: string) =>
  url.replace(/\.(jpg|JPG)$/, (ext) => `${suffix}${ext}`);

/**
 * Rovers shoot in stereo: the same instant through a left and a right eye, and
 * both eyes are published as separate frames. A sol of 214 Perseverance frames
 * is really about 149 scenes. Keep the left eye when a right one shares its
 * capture time and its camera differs only by which eye it is.
 *
 * Measured on Perseverance sol 1972: 214 frames in, 149 out, and Mastcam-Z
 * drops from 71+71 to 41+41. On Curiosity the two eyes rarely share a
 * timestamp, so this removes almost nothing there — which is correct, they are
 * genuinely separate exposures.
 */
function dropStereoTwins(frames: FrameThumb[]): FrameThumb[] {
  const seen = new Map<string, number>();
  const out: FrameThumb[] = [];
  for (const f of frames) {
    const base = f.instrument.replace(/_(LEFT|RIGHT)/, '');
    const key = `${f.capturedUtc}|${base}`;
    const at = seen.get(key);
    if (at === undefined) { seen.set(key, out.length); out.push(f); continue; }
    if (/LEFT/.test(f.instrument) && /RIGHT/.test(out[at]!.instrument)) out[at] = f;
  }
  return out;
}

/**
 * How much a camera is worth opening a post on, highest first. The raw feeds
 * carry no "featured" or "interesting" flag; the closest thing NASA publishes
 * is MSL's `instrument_sort` (Mastcam 1, ChemCam RMI 4, Navcam 7-8), and this
 * ordering matches it while also covering Perseverance, whose feed has no
 * equivalent field.
 */
const CAMERA_RANK: [RegExp, number][] = [
  [/MCZ|MAST|ZCAM/, 0],        // colour, scenic
  [/NAVCAM|NAV_/, 1],          // wide, grey
  [/RMI|SUPERCAM|CHEMCAM/, 2], // distant detail
  [/HAZ/, 3],                  // wheels and ground
];
export function openingFrame(frames: { instrument: string }[]): number {
  let best = 0;
  let bestRank = 99;
  for (let i = 0; i < frames.length; i++) {
    const inst = frames[i]!.instrument;
    const rank = CAMERA_RANK.find(([re]) => re.test(inst))?.[1] ?? 4;
    if (rank < bestRank) { bestRank = rank; best = i; if (rank === 0) break; }
  }
  return best;
}

export interface SolImages {
  sol: number;
  count: number;      // total frames that sol (may exceed the loaded sample)
  frames: FrameThumb[];
  more: string;       // deep link to the full set on mars.nasa.gov
}

const cache = new Map<string, SolImages>();

// ---- newest published frames ------------------------------------------------
// The bundled snapshot is only as fresh as the last data refresh, and NASA
// publishes in bursts through the day, so the feed asks for the most recently
// *published* frames (ordered by date_received) rather than guessing a sol.
const latestCache = new Map<string, { at: number; frames: FrameThumb[] }>();
const LATEST_TTL_MS = 5 * 60 * 1000;

/**
 * `force` skips the cache on the way in, never on the way out. A poll that
 * reads a five-minute-old answer defeats the point of polling, and the sol a
 * rover is filling right now keeps growing under the same key.
 */
export async function fetchLatestFrames(
  roverId: string,
  limit = 48,
  force = false,
): Promise<FrameThumb[]> {
  const key = `${roverId}:${limit}`;
  const hit = latestCache.get(key);
  if (!force && hit && Date.now() - hit.at < LATEST_TTL_MS) return hit.frames;

  const url =
    roverId === 'curiosity'
      ? `https://mars.nasa.gov/api/v1/raw_image_items/?order=date_received+desc&per_page=${limit * 2}&page=0&condition_1=msl%3Amission`
      : `https://mars.nasa.gov/rss/api/?feed=raw_images&category=mars2020&feedtype=json&num=${limit}&page=0&order=date_received+desc`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`${roverId} latest: HTTP ${res.status}`);
  const data = await res.json();

  const frames: FrameThumb[] =
    roverId === 'curiosity'
      ? ((data.items ?? []) as any[])
          .filter((im) => !im.is_thumbnail && im.url)
          .map((im) => ({
            file: mslVariant(im.url, '-thm'),
            view: mslVariant(im.url, '-br'),
            full: im.url,
            sourceUrl: im.url,
            instrument: im.instrument ?? 'CAMERA',
            capturedUtc: im.date_taken ?? '',
            receivedUtc: im.date_received ?? '',
            sol: im.sol ?? null,
          }))
          .slice(0, limit)
      : ((data.images ?? []) as any[])
          .map((im) => {
            const f = im.image_files ?? {};
            return {
              file: f.small ?? f.medium ?? f.large ?? f.full_res,
              view: f.medium ?? f.large ?? f.small,
              full: f.large ?? f.full_res ?? f.medium ?? f.small,
              sourceUrl: im.link ?? f.full_res ?? f.large,
              instrument: im.camera?.instrument ?? 'CAMERA',
              capturedUtc: im.date_taken_utc ?? '',
              receivedUtc: im.date_received ?? '',
              sol: im.sol ?? null,
            };
          })
          .filter((f) => f.file);

  const deduped = dropStereoTwins(frames);
  latestCache.set(key, { at: Date.now(), frames: deduped });
  return deduped;
}

export async function fetchSolImages(
  roverId: string,
  sol: number,
  limit = 120,
  force = false,
): Promise<SolImages> {
  const key = `${roverId}:${sol}:${limit}`;
  const hit = cache.get(key);
  if (!force && hit) return hit;

  const result =
    roverId === 'curiosity'
      ? await fetchCuriosity(sol, limit)
      : await fetchPerseverance(sol, limit);
  cache.set(key, result);
  return result;
}

async function fetchPerseverance(sol: number, limit: number): Promise<SolImages> {
  const url =
    `https://mars.nasa.gov/rss/api/?feed=raw_images&category=mars2020&feedtype=json` +
    `&num=${limit}&page=0&order=sol+desc&sol=${sol}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`mars2020 sol ${sol}: HTTP ${res.status}`);
  const data = await res.json();
  const imgs: any[] = data.images ?? [];
  const frames: FrameThumb[] = imgs.map((im) => {
    const f = im.image_files ?? {};
    return {
      file: f.small ?? f.medium ?? f.large ?? f.full_res,
      view: f.medium ?? f.large ?? f.small,
      full: f.large ?? f.full_res ?? f.medium ?? f.small,
      sourceUrl: im.link ?? f.full_res ?? f.large,
      instrument: im.camera?.instrument ?? 'CAMERA',
      capturedUtc: im.date_taken_utc ?? im.date_taken_mars ?? '',
      receivedUtc: im.date_received ?? '',
      sol: im.sol ?? sol,
    };
  }).filter((f) => f.file);
  return {
    sol,
    count: data.num_images ?? frames.length,
    frames: dropStereoTwins(frames).slice(0, limit), // the feed ignores `num` when filtering by sol
    more: `https://mars.nasa.gov/mars2020/multimedia/raw-images/?order=sol+desc&per_page=100&page=0&begin_sol=${sol}&end_sol=${sol}`,
  };
}

async function fetchCuriosity(sol: number, limit: number): Promise<SolImages> {
  // Each full frame has a low-res `is_thumbnail` twin, so over-fetch and drop
  // the thumbnails before trimming to `limit` full-resolution frames.
  const url =
    `https://mars.nasa.gov/api/v1/raw_image_items/?order=sol+desc&per_page=${limit * 2}&page=0` +
    `&condition_1=msl%3Amission&condition_2=${sol}%3Asol%3Agte&condition_3=${sol}%3Asol%3Alte`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`msl sol ${sol}: HTTP ${res.status}`);
  const data = await res.json();
  const items: any[] = (data.items ?? []).filter((im: any) => !im.is_thumbnail);
  const frames: FrameThumb[] = items.map((im) => ({
    file: mslVariant(im.url, '-thm'),
    view: mslVariant(im.url, '-br'),
    full: im.url,
    sourceUrl: im.url,
    instrument: im.instrument ?? 'CAMERA',
    capturedUtc: im.date_taken ?? '',
    receivedUtc: im.date_received ?? '',
    sol: im.sol ?? sol,
  })).filter((f) => f.file).slice(0, limit);
  return {
    sol,
    count: items.length,
    frames: dropStereoTwins(frames),
    more: `https://mars.nasa.gov/msl/multimedia/raw-images/?order=sol+desc&per_page=100&page=0&begin_sol=${sol}&end_sol=${sol}`,
  };
}
